pilot/pkg/xds/internalgen.go

```diff
@@ -142,14 +155,111 @@ func (sg *InternalGen) startPush(typeURL string, data []proto.Message) {
 // We can also expose ACKS.
 func (sg *InternalGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
 	res := []*any.Any{}
-	switch w.TypeUrl {
+
+	// Look for ?query parameters.  We don't use url.Parse() because we have no scheme
+	typeURL := w.TypeUrl
+	var qparams url.Values
+	if qIndex := strings.Index(w.TypeUrl, "?"); qIndex >= 0 {
+		typeURL = w.TypeUrl[:qIndex]
+		var err error
+		qparams, err = url.ParseQuery(w.TypeUrl[qIndex+1:])
+		if err != nil {
+			adsLog.Infof("Invalid TypeUrl query params: %q", w.TypeUrl)
+		}
+	}
+
+	switch typeURL {
 	case TypeURLConnections:
 		sg.Server.adsClientsMutex.RLock()
 		// Create a temp map to avoid locking the add/remove
 		for _, v := range sg.Server.adsClients {
 			res = append(res, util.MessageToAny(v.xdsNode))
 		}
 		sg.Server.adsClientsMutex.RUnlock()
+	case TypeDebugSyncronization:
+		res = sg.debugSyncz()
+	case TypeDebugConfigDump:
+		proxyID := qparams.Get("proxyID")
+		if proxyID == "" {
+			adsLog.Info("Config Dump w/o proxyID query parameter")
+			break
+		}
+		res = sg.debugConfigDump(proxyID)
+	default:
+		adsLog.Infof("Unknown TypeUrl: %q", w.TypeUrl)
+	}
+	return res
+}
+
+func (sg *InternalGen) debugSyncz() []*any.Any {
+	res := []*any.Any{}
+
+	sg.Server.adsClientsMutex.RLock()
+	for _, con := range sg.Server.adsClients {
+		con.mu.RLock()
+		// Skip "nodes" without metdata (they are probably istioctl queries!)
+		if con.node != nil && con.node.Metadata != nil && con.node.Metadata.ProxyConfig != nil {
+			xdsConfigs := []*status.PerXdsConfig{}
+			for stype, watchedResource := range con.node.Active {
+				pxc := &status.PerXdsConfig{
+					Status: debugSyncStatus(watchedResource),
+				}
+				switch stype {
+				case pilot_xds_v3.ListenerShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ListenerConfig{}
+				case pilot_xds_v3.RouteShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_RouteConfig{}
+				case pilot_xds_v3.EndpointShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_EndpointConfig{}
+				case pilot_xds_v3.ClusterShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ClusterConfig{}
+				default:
+					adsLog.Warnf("Unknown watchedResource type: %q", stype)
+				}
+				xdsConfigs = append(xdsConfigs, pxc)
+			}
+			clientConfig := &status.ClientConfig{
+				Node: &core.Node{
+					Id: con.node.ID,
+				},
+				XdsConfig: xdsConfigs,
+			}
+			log.Infof("Trying to show status of %#v\n", clientConfig)
+			res = append(res, util.MessageToAny(clientConfig))
+		}
+		con.mu.RUnlock()
 	}
+	sg.Server.adsClientsMutex.RUnlock()
+
 	return res
 }
+
+func debugSyncStatus(wr *model.WatchedResource) status.ConfigStatus {
+	if wr.NonceSent == "" {
+		return status.ConfigStatus_NOT_SENT
```

Comment:
Is it possible to add ConfigStatus_ERROR here to represent NACKs?

Reply:
The logic comes from https://github.com/istio/istio/blob/master/istioctl/pkg/writer/pilot/status.go#L106 .  In 1.6 we don't use ERROR.  @liamawhite wrote the original code.  @liamawhite I have questions

- @therealmitchconnors feels we should return ERROR instead of STALE.  What do you think?
- The 1.6 code has the possibility of returning `STALE (Never Acknowledged)`.  I have not seen that code in years but perhaps it happens in the brief window when Istiod is pushing on another thread?

Reply:
Stale and NACK are different things, I'm not sure @therealmitchconnors was asking for that though? There a couple of possibilities:

- Sent but not yet ack'd > threshold -> `STALE`
- Nacked -> `ERROR`
- Ack'd `SYNCED`

I think also `NOT_SENT` should change to something like `OK` or just blank. Users get confused and think not sent indicates and error but it's actually fine.